jsoniter/src/io/github/iltotore/iron/jsoniter.scala

```diff
@@ -0,0 +1,17 @@
+package io.github.iltotore.iron
+
+import com.github.plokhotnyuk.jsoniter_scala.core.*
+import com.github.plokhotnyuk.jsoniter_scala.macros.*
+
+object jsoniter:
+
+  /**
+   * Creates a [[JsonValueCodec]] for refined types
+   *
+   * @param constraint the [[Constraint]] implementation to test the decoded value.
+   */
+  inline given makeCodec[A, B](using inline constraint: Constraint[A, B]): JsonValueCodec[A :| B] = new:
+    private val codec = JsonCodecMaker.make[A]
+    def decodeValue(in: JsonReader, default: A :| B): A :| B = codec.decodeValue(in, default).refineEither[B].fold(in.decodeError, identity)
```

Comment:
This is causing an unnecessary boxing of the primitives. A potentially better approach would be to do this instead:

```scala
  inline given makeCodec[A, B](using inline constraint: Constraint[A, B]): JsonValueCodec[A :| B] = new:
    private val codec = JsonCodecMaker.make[A]
    def decodeValue(in: JsonReader, default: A :| B): A :| B =
      val decoded = codec.decodeValue(in, default)
      if constraint.test(decoded) then decoded.asInstanceOf[A :| B]
      else in.decodeError(constraint.message)
    def encodeValue(x: A :| B, out: JsonWriter): Unit = codec.encodeValue(x, out)
    def nullValue: A :| B = null.asInstanceOf[A :| B]
```

I'll need to look into this a bit more to see which approach might be better

Reply:
This seems to be a better approach to me.